[BD-6] OEP-18 compliance

morenol opened the pull request:
Add changes to make this repo OEP-18 compliant. 

I moved the constraints to constraints.txt, I think that we can unpin some of them but that can be done in other PR.

## Reviewers
- [x] @awais786
- [x] @ericfab179

edx-webhook commented:
Thanks for the pull request, @morenol! I've created [OSPR-4612](https://openedx.atlassian.net/browse/OSPR-4612) to keep track of it in JIRA. JIRA is a place for product owners to prioritize feature reviews by the engineering development teams.

Feel free to add as much of the following information to the ticket:
- supporting documentation
- edx-code email threads
- timeline information ("this must be merged by XX date", and why that is)
- partner information ("this is a course on edx.org")
- any other information that can help Product understand the context for the PR

All technical communication about the code itself will still be done via the GitHub pull request interface. As a reminder, [our process documentation is here](http://edx-developer-guide.readthedocs.org/en/latest/process/overview.html).

mzulqarnain1 commented on requirements/base.in:
```diff
@@ -0,0 +1,19 @@
+-c constraints.txt
+
+attrs<19.0.0
```
We should use constraints.txt for this purpose.

awais786 commented:
@ormsbee could you review this one?

symbolist commented on requirements/base.in:
```diff
@@ -0,0 +1,19 @@
+django
+django-cors-headers
+django-environ
+django-extensions
```
It does not look like we are using it? It needs to be added to `INSTALLED_APPS` to work?

https://django-extensions.readthedocs.io/en/latest/installation_instructions.html#configuration

morenol replied:
Removed, it seems that it is not being used.

morenol commented on requirements/base.in:
```diff
@@ -0,0 +1,19 @@
+djangorestframework
+djangorestframework-expander
+django-filter
+git+https://github.com/alanjds/drf-nested-routers.git
```
The last PyPi release of that drf-nested-routers was in 2018 and it does not have support for Django2.2.

I made a PR some weeks ago in order to help the maintainer of that repo to create a new release, but I have not received any update yet.

https://github.com/alanjds/drf-nested-routers/pull/175

What do you think if I change this with 

git+https://github.com/alanjds/drf-nested-routers.git@1943ab761cb3b6b0b3128d5be409571d3d83e64a#egg=drf-nested-routers

just like the OEP18 specification suggests?

symbolist replied:
Thanks, that should work. Can you also specify the version number at the end. It makes it clearer what version is installed without needing to go check out the relevant repo?

```git+https://github.com/edx/django-ratelimit-backend.git@v2.0.1a5#egg=django-ratelimit-backend==2.0.1a5```

morenol replied:
Do you mean use an arbitrary version at the end?
 Something like git+https://github.com/alanjds/drf-nested-routers.git@1943ab761cb3b6b0b3128d5be409571d3d83e64a#egg=drf-nested-routers==0.91a (the last version in that repo is 0.91)

symbolist replied:
I don't think we need to make it arbitrary. My understanding is that the hash will always be used for selecting the code version. If the package version the commit is at is `0.91` we can go with:

```
git+https://github.com/alanjds/drf-nested-routers.git@1943ab761cb3b6b0b3128d5be409571d3d83e64a#egg=drf-nested-routers==0.91
```

https://open-edx-proposals.readthedocs.io/en/latest/oep-0018-bp-python-dependencies.html#installing-dependencies-from-urls

alanjds replied:
Just to inform. The https://github.com/alanjds/drf-nested-routers/pull/186 replaces alanjds/drf-nested-routers#175 and was merged.

morenol commented on requirements/optional.in:
```diff
@@ -0,0 +1,3 @@
+-c constraints.txt
+
+newrelic
```
No, I just added that in order to keep everything working as it was before these changes. Should I remove that from production or from optional?

symbolist replied:
I looked at some of the other repos (e.g. https://github.com/edx/ecommerce/blob/master/requirements/production.in) and it looks like this is being added to `production.in`. So lets just remove `optional.in` altogether.

symbolist commented:
@morenol Thanks! Almost there. 🙂

symbolist commented:
@morenol Thanks, please update the drf-nested-routers url and then squash all the commits.

symbolist commented:
@morenol Can you please take a look at the failing tests? Once they are fixed I'll merge this. Thanks.

morenol commented:
> @morenol Can you please take a look at the failing tests? Once they are fixed I'll merge this. Thanks.

I think that the test errors are also in the master branch, I am going to try to fix them in this PR, if they are simple, otherwise I will create a separate PR to fix them.

morenol commented:
@symbolist I fixed the tests. I also added constraint due to the django1.11 usage and due to an [error](https://github.com/FactoryBoy/factory_boy/issues/780) in the installation of factory_boy last version.

I guess that if we merge this we would need to remove the constraints again in the PR with the django2.2 changes

symbolist reviewed (approved):
Thanks for this update! Since this PR got started first, we can merge it first too. :)
